src/chunk_constraint.c

```diff
@@ -312,46 +424,98 @@ chunk_constraint_create(const ChunkConstraint *cc, Oid chunk_oid, int32 chunk_id
 	if (!OidIsValid(chunk_constraint_oid))
 		return InvalidOid;
 
-	if (!is_dimension_constraint(cc))
-	{
-		Oid hypertable_constraint_oid =
-			get_relation_constraint_oid(hypertable_oid,
-										NameStr(cc->fd.hypertable_constraint_name),
-										false);
-		HeapTuple tuple = SearchSysCache1(CONSTROID, hypertable_constraint_oid);
+	Oid hypertable_constraint_oid =
+		get_relation_constraint_oid(hypertable_oid,
+									NameStr(cc->fd.hypertable_constraint_name),
+									false);
+	HeapTuple tuple = SearchSysCache1(CONSTROID, hypertable_constraint_oid);
 
-		if (HeapTupleIsValid(tuple))
-		{
-			FormData_pg_constraint *constr = (FormData_pg_constraint *) GETSTRUCT(tuple);
+	if (HeapTupleIsValid(tuple))
+	{
+		FormData_pg_constraint *constr = (FormData_pg_constraint *) GETSTRUCT(tuple);
 
-			if (OidIsValid(constr->conindid) && constr->contype != CONSTRAINT_FOREIGN)
-				ts_chunk_index_create_from_constraint(hypertable_id,
-													  hypertable_constraint_oid,
-													  chunk_id,
-													  chunk_constraint_oid);
+		if (OidIsValid(constr->conindid) && constr->contype != CONSTRAINT_FOREIGN)
+			ts_chunk_index_create_from_constraint(hypertable_id,
+												  hypertable_constraint_oid,
+												  chunk_id,
+												  chunk_constraint_oid);
 
-			ReleaseSysCache(tuple);
-		}
+		ReleaseSysCache(tuple);
 	}
 
 	return chunk_constraint_oid;
 }
 
+static const DimensionSlice *
+get_slice_with_id(const Hypercube *cube, int32 id)
+{
+	int i;
+
+	for (i = 0; i < cube->num_slices; i++)
+	{
+		const DimensionSlice *slice = cube->slices[i];
+
+		if (slice->fd.id == id)
+			return slice;
+	}
+
+	return NULL;
+}
+
 /*
  * Create a set of constraints on a chunk table.
  */
 void
-ts_chunk_constraints_create(const ChunkConstraints *ccs, Oid chunk_oid, int32 chunk_id,
-							Oid hypertable_oid, int32 hypertable_id)
+ts_chunk_constraints_create(const Hypertable *ht, const Chunk *chunk)
 {
+	const ChunkConstraints *ccs = chunk->constraints;
+	List *newconstrs = NIL;
 	int i;
 
 	for (i = 0; i < ccs->num_constraints; i++)
-		chunk_constraint_create(&ccs->constraints[i],
-								chunk_oid,
-								chunk_id,
-								hypertable_oid,
-								hypertable_id);
+	{
+		const ChunkConstraint *cc = &ccs->constraints[i];
+
+		if (is_dimension_constraint(cc))
+		{
+			const DimensionSlice *slice = get_slice_with_id(chunk->cube, cc->fd.dimension_slice_id);
+			const Dimension *dim;
+			Constraint *constr;
+
+			dim = ts_hyperspace_get_dimension_by_id(ht->space, slice->fd.dimension_id);
```

Comment:
Variable `slice` can in theory be NULL here if there is something wrong with the metadata. You might want to ensure that it is not NULL before starting to de-reference it.

```suggestion
			Ensure(slice, "dimension slice was missing");
			dim = ts_hyperspace_get_dimension_by_id(ht->space, slice->fd.dimension_id);
```